index/index.go

```diff
@@ -1025,11 +1064,21 @@ type Decoder struct {
 }
 
 // Postings returns a postings list for b and its number of elements.
-func (dec *Decoder) Postings(b []byte) (int, Postings, error) {
+func (dec *Decoder) Postings(b []byte, reusePos Postings) (int, Postings, error) {
 	d := encoding.Decbuf{B: b}
 	n := d.Be32int()
-	l := d.Get()
-	return n, newBigEndianPostings(l), d.Err()
+	if bep, ok := reusePos.(*bigEndianPostings); ok {
```

Comment:
I have this special case even though I am calling `reusePos.Reset(...)` later because `l := d.Get()` needs an alloc as we are passing it as a pointer. We can skip that alloc if `reusePos` is `bigEndianPostings`.

This looks like a hack but doesn't take away the generic interface flow.

Reply:
If the ONLY gain for any caller of this all `Postings` implementation is if the `Postings` is  `(*bigEndianPostings)` this makes the interface quite complex. Maybe we should expect `*BigEndianPostings` then explicitly? This makes `Postings` quite complext to use.

Reply:
This would mean assuming the inner details of `IndexReader` interface. It could also become harder to refactor if a specific posting is mentioned in the function but the index uses multiple postings (Example if something like #629 is merged). Currently, it would mean modifying this function rather than modifying function signatures if such changes are added in the future.

But I agree that it might become harder to maintain this. I am still trying to figure out any better ways to do it (don't want to let it go as gains are significant).